language/compiler/ir-to-bytecode/syntax/src/lexer.rs

```diff
@@ -311,41 +132,248 @@ impl<'input, 'builder> Lexer<'input, 'builder> {
         let text = self.text.trim_start();
         let whitespace = self.text.len() - text.len();
         let start_offset = self.consumed + whitespace;
-        if text.is_empty() {
-            self.text = text;
-            self.consumed = start_offset;
-            self.token = (start_offset, Token(Tok::EOF, ""), start_offset);
-            Ok(())
-        } else {
-            let mut longest_match = 0;
-            let mut index: Option<usize> = None;
-            for (i, token_exp) in self.regex_vec.iter().enumerate() {
-                if let Some(m) = token_exp.find(text) {
-                    let len = m.end();
-                    if len >= longest_match {
-                        longest_match = len;
-                        index = Some(i.try_into().unwrap());
+        let c: char = match text.chars().next() {
+            Some(next_char) => next_char,
+            None => {
+                self.text = text;
+                self.consumed = start_offset;
+                self.token = (start_offset, Token(Tok::EOF, ""), start_offset);
+                return Ok(());
+            }
+        };
+        let (tok, len) = match c {
+            '0'..='9' => {
+                if (text.starts_with("0x") || text.starts_with("0X")) && text.len() > 2 {
+                    let hex_len = get_hex_digits_len(&text[2..]);
+                    if hex_len == 0 {
+                        // Fall back to treating this as a "0" token.
+                        (Tok::U64Value, 1)
+                    } else {
+                        (Tok::AccountAddressValue, 2 + hex_len)
+                    }
+                } else {
+                    (Tok::U64Value, get_decimal_digits_len(&text))
+                }
+            }
+            'a'..='z' | 'A'..='Z' | '$' | '_' => {
+                let len = get_name_len(&text);
+                let name = &text[..len];
+                match &text[len..].chars().next() {
+                    Some('"') => {
+                        // Special case for ByteArrayValue: h\"[0-9A-Fa-f]*\"
+                        let bvlen = get_byte_array_value_len(&text[(len + 1)..]);
+                        if name == "h" && bvlen > 0 {
+                            (Tok::ByteArrayValue, 2 + bvlen)
+                        } else {
+                            (get_name_token(name), len)
+                        }
```

Comment:
should we case on "h" first before doing this check for `get_byte_array_value_len`? 
i.e. 

```
if name == "h" {
    match &text[len..].chars().next() {
        Some('"') => return ....,
        _ => (),
    }
}
```

Reply:
OK. (Todd helped me figure out how to write the code that I wanted in Rust.)